File: 백준/Gold/16236. 아기 상어/아기 상어.cc
#include<iostream>
#include<algorithm>
#include<queue>
#include<vector>
using namespace std;

class Shark{
private:
    int N, size, eating, nowX, nowY, line;
    int **map;
    bool **visited;
public:
    Shark(int n){
        N = n;
        size = 2, eating = 0;
    }
    void input();
    bool bfs(int x, int y);
    void answer();
};

void Shark::input(){
    int temp;
    map = new int*[N];
    for(int i = 0; i < N; i++){
        map[i] = new int[N];
        for(int j = 0; j < N; j++){
            cin >> temp;
            map[i][j] = temp;
            if(temp == 9){
                nowX = i, nowY = j;
            }
        }
    }
}

bool Shark::bfs(int x, int y){
    int firstX = x, firstY = y;
    visited = new bool*[N];
    for(int i = 0; i < N; i++){
        visited[i] = new bool[N]{false};
    }
    int dx[4] = {-1,0,0,1};
    int dy[4] = {0,-1,1,0};
    queue<pair<int,pair<int,int>>> q; //거리,좌표 q에 저장
    priority_queue<pair<int,int>,vector<pair<int,int>>,greater<pair<int,int>>> pq;
    //이동가능한 위치들 우선순위 큐에 저장->오름차순이라 맨앞이 가장 위면서 가장 왼쪽 좌표
    q.push(make_pair(0,make_pair(x,y)));
    visited[x][y] = true;
    map[x][y] = 0;
    int temp = -1;
    while(!q.empty()){
        int nX = q.front().second.first;
        int nY = q.front().second.second;
        int dis = q.front().first;
        if(dis == temp){
            break;
        }
        q.pop();
        for(int i = 0; i < 4; i++){
            int nextX = nX + dx[i];
            int nextY = nY + dy[i];
            if(nextX >= 0 && nextX < N && nextY >= 0 && nextY < N && map[nextX][nextY] <= size && !visited[nextX][nextY]){
                if(map[nextX][nextY] == 0 || map[nextX][nextY] == size){
                    q.push(make_pair(dis+1,make_pair(nextX,nextY)));
                    visited[nextX][nextY] = true;
                }
                else{
                    visited[nextX][nextY] = true;
                    if(temp == -1){
                        temp = dis+1;
                    }
                    pq.push(make_pair(nextX,nextY));
                    q.push(make_pair(dis+1,make_pair(nextX,nextY)));
                }
            }
        }
    }
    if(pq.size() == 0){
        return false;
    }
    else{
        eating += 1;
        if(size == eating){
            eating = 0;
            size += 1;
        }
        nowX = pq.top().first, nowY = pq.top().second;
        line = temp;
        return true;
    }
}

void Shark::answer(){
    int answerNum = 0;
    while(bfs(nowX,nowY)){
        answerNum += line;
    }
    cout << answerNum << "\n";
}

int main(){
    cin.tie(NULL); cout.tie(NULL);
    ios::sync_with_stdio(false);
    int inputNum;
    cin >> inputNum;
    Shark s(inputNum);
    s.input();
    s.answer();
    return 0;
}

File: 백준/Gold/16236. 아기 상어/아기 상어.py
import sys
input = sys.stdin.readline
import copy

n = int(input())
mainSpace = []
visited = []
dx = [-1,0,0,1]
dy = [0,-1,1,0]
for i in range(n):
    temp = list(map(int,input().split()))
    mainSpace.append([])
    visited.append([False]*n)
    for j in range(n):
        if temp[j] == 9:
            shark = (i,j)
            visited[i][j] = True
            temp[j] = 0
        mainSpace[i].append(temp[j])

size = 2
eat = 0
day = 0
while True:
    space = copy.deepcopy(mainSpace)
    q = []
    q.append(shark)
    search = False
    length = 1
    use = 0
    tempD = 0
    searchq = []
    while q:
        nowX, nowY = q.pop(0)
        space[nowX][nowY] = -1
        for i in range(4):
            nextX = nowX + dx[i]
            nextY = nowY + dy[i]
            if nextX >= 0 and nextX < n and nextY < n and nextY >= 0:
                if space[nextX][nextY] > size:
                    continue
                if space[nextX][nextY] == 0 or space[nextX][nextY] == size:
                    q.append((nextX,nextY))
                    space[nextX][nextY] = -1
                elif 0 < space[nextX][nextY] < size and visited[nextX][nextY] != True:
                    searchq.append((nextX,nextY))
                    search = True
        use += 1
        if use == length:
            if search:
                nextX,nextY = min(searchq)
                visited[nextX][nextY] = True
                mainSpace[nextX][nextY] = 0
                eat += 1
                if eat == size:
                    eat = 0
                    size += 1
                day += 1 + tempD
                shark = (nextX,nextY)
                break
            tempD += 1
            use = 0
            length = len(q)
            q.sort()
    if not search:
        break
print(day)